OSDOCS-10818: Introducing oc adm upgrade status (tech preview) (#77273)
openshift/openshift-docs: modules/update-upgrading-oc-adm-upgrade-status.adoc

// Module included in the following assemblies:
//
// * updating/updating_a_cluster/updating-cluster-cli.adoc

:_mod-docs-content-type: CONTEXT
[id="update-upgrading-oc-adm-upgrade-status_{context}"]
= Using `oc adm upgrade status` to retrieve information about a cluster update (Technology Preview)

> When updating your cluster, it is useful to understand how your update is progressing. While the `oc adm upgrade` command returns limited information about the status of your update, this release introduces the `oc adm upgrade status` command as a Technology Preview feature. This command decouples status information from the `oc adm upgrade` command and provides specific information regarding a cluster update, including the status of the control plane and worker node updates.

Reviewer: One thing we probably need to stress out is that because this command is a Technology Preview, it has to be explicitly enabled by setting the `OC_ENABLE_CMD_UPGRADE_STATUS` environmental variable to `true`, otherwise it will not work and oc will behave as if no such command exists.

> The `oc adm upgrade status` command only returns information on the status of your current cluster update. It does not provide read or write capabilities and will not alter any information in your Operator. 

Reviewer: "It does not provide read or write capabilities" sounds weird. Any command does not provide read/write capabilities. The command *reads* the state of the cluster but never writes or modifies anything. That's probably what the second half of the sentence says, but IMO "information in your Operator" is not precise, I'd say something like "never alters any state in your cluster" (I guess "operator" slipped in by mistake).

The `oc adm upgrade status` command is read-only and will never alter any state in your cluster. 

[IMPORTANT]
====	
The `oc adm upgrade status` command is a Technology Preview feature only. Technology Preview features are not supported with Red Hat production service level agreements (SLAs) and might not be functionally complete. Red Hat does not recommend using them in production. These features provide early access to upcoming product features, enabling customers to test functionality and provide feedback during the development process.

For more information about the support scope of Red Hat Technology Preview features, see link:https://access.redhat.com/support/offerings/techpreview[Technology Preview Features - Scope of Support].

> ====

Reviewer 2: You can use the [TP snippet](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#indicating-technology-preview-features) here.

The `oc adm upgrade status` command can be used for clusters from version 4.12 up to the latest supported release. 

[IMPORTANT]
====

> While your cluster does not need to be a TechPreview-enabled cluster, you must enable the `oc adm upgrade status` Technology Preview environment variable, otherwise the `oc` will not recognize the command and you will not be able to use the feature. 

Reviewer 2: ```suggestion
While your cluster does not need to be a Technology Preview-enabled cluster, you must enable the `oc adm upgrade status` Technology Preview environment variable, otherwise the `oc` will not recognize the command and you will not be able to use the feature. 
```

====

.Procedure 

> * To enable the `oc adm upgrade status` Technology Preview environment variable, set the `OC_ENABLE_CMD_UPGRADE_STATUS` environmental variable to `true` by running the following command:

Reviewer 2: Does the environment variable need to be passed with the command every time it is run (or maybe that's just how we'd prefer to show it being run for now?), or can you set the environment variable globally and then just run `oc adm upgrade status` by itself?

If the former, I wonder if the following would be more accurate as far as what running this command is fully doing; that is, that it's both setting the variable (for just this run?) and actually running the command at the same time (and this would also match the procedure module's heading more closely):
```suggestion
* To retrieve information about a cluster update with the `oc adm upgrade status` command, enable the `oc adm upgrade status` Technology Preview environment variable by setting `OC_ENABLE_CMD_UPGRADE_STATUS` to `true` while running the command:
```
If the latter, an alternative version of this could maybe be a step 1 shows setting `OC_ENABLE_CMD_UPGRADE_STATUS=true` globally, and step 2 just runs `oc adm upgrade status` by itself, followed by some example output.

Reviewer: > Does the environment variable need to be passed with the command every time it is run (or maybe that's just how we'd prefer to show it being run for now?), or can you set the environment variable globally and then just run oc adm upgrade status by itself?

You can definitely set it globally and then run the command alone. Whether you do so or pass it for a single execution is simply a user preference. I agree that maybe presenting it as two separate steps may be better.

Reviewer 3: A benefit to doc'ing as one line is that it's easier to narrowly copy/paste, without having to scroll up and see that you also need to export the environment variable in a separate command.  But :shrug:, I dunno if convenient copy-pasting is worth the trouble of trying to explain the one-line form, vs. spreading out into two separate commands.

> * To enable the `oc adm upgrade status` Technology Preview environment variable, set the `OC_ENABLE_CMD_UPGRADE_STATUS` environmental variable to `true` by running the following command:
> 
> [source,terminal]

Reviewer 2: Attach the command to the step:
```suggestion
* To enable the `oc adm upgrade status` Technology Preview environment variable, set the `OC_ENABLE_CMD_UPGRADE_STATUS` environmental variable to `true` by running the following command:
+
[source,terminal]
```

----
$ OC_ENABLE_CMD_UPGRADE_STATUS=true ./oc adm upgrade status

> ----
> 
> The following output is an example of an update that is progressing successfully:
> 
> .Example ouput

Reviewer 2: Attach the example output to the step that runs the command (also note typo):
```suggestion
----
+
The following output is an example of an update that is progressing successfully:
+
.Example output
```

> [source,terminal]

Reviewer 2: Consider a collapsible block for this output since it's kind of a long-o: https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#collapsible-content

ip-10-0-92-180.us-east-2.compute.internal   Outdated      Pending    4.14.0    ?

= Worker Upgrade =

= Worker Pool =
Worker Pool:     worker
Assessment:      Progressing
Completion:      0%
Worker Status:   3 Total, 2 Available, 1 Progressing, 3 Outdated, 1 Draining, 0 Excluded, 0 Degraded

Worker Pool Nodes
NAME                                        ASSESSMENT    PHASE      VERSION   EST    MESSAGE
ip-10-0-4-159.us-east-2.compute.internal    Progressing   Draining   4.14.0    +10m
ip-10-0-20-162.us-east-2.compute.internal   Outdated      Pending    4.14.0    ?
ip-10-0-99-40.us-east-2.compute.internal    Outdated      Pending    4.14.0    ?

= Worker Pool =
Worker Pool:     infra
Assessment:      Progressing
Completion:      0%
Worker Status:   1 Total, 0 Available, 1 Progressing, 1 Outdated, 1 Draining, 0 Excluded, 0 Degraded

Worker Pool Node
NAME                                             ASSESSMENT    PHASE      VERSION   EST    MESSAGE
ip-10-0-4-159-infra.us-east-2.compute.internal   Progressing   Draining   4.14.0    +10m

= Update Health =
SINCE   LEVEL   IMPACT   MESSAGE
14m4s   Info    None     Update is proceeding well
----

> With this information, you can made informed decisions on how to proceed with your update.

Reviewer 2: ```suggestion
With this information, you can make informed decisions on how to proceed with your update.
```

// Module included in the following assemblies:
//
// * updating/updating_a_cluster/updating-cluster-cli.adoc

:_mod-docs-content-type: PROCEDURE
[id="update-upgrading-oc-adm-upgrade-status_{context}"]
= Retrieving information about a cluster update using oc adm upgrade status (Technology Preview)

When updating your cluster, it is useful to understand how your update is progressing. While the `oc adm upgrade` command returns limited information about the status of your update, this release introduces the `oc adm upgrade status` command as a Technology Preview feature. This command decouples status information from the `oc adm upgrade` command and provides specific information regarding a cluster update, including the status of the control plane and worker node updates.

The `oc adm upgrade status` command is read-only and will never alter any state in your cluster. 

> :FeatureName: The oc adm upgrade status command

Reviewer 2: ```suggestion
:FeatureName: The `oc adm upgrade status` command
```

include::snippets/technology-preview.adoc[]

The `oc adm upgrade status` command can be used for clusters from version 4.12 up to the latest supported release. 

[IMPORTANT]
====

> While your cluster does not need to be a Technology Preview-enabled cluster, you must enable the `oc adm upgrade status` Technology Preview environment variable, otherwise the `oc` will not recognize the command and you will not be able to use the feature. 

Reviewer 2: Per https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/term_glossary.adoc#openshift-cli:
```suggestion
While your cluster does not need to be a Technology Preview-enabled cluster, you must enable the `oc adm upgrade status` Technology Preview environment variable, otherwise the OpenShift CLI (`oc`) will not recognize the command and you will not be able to use the feature. 
```
or using the common attribute:
```suggestion
While your cluster does not need to be a Technology Preview-enabled cluster, you must enable the `oc adm upgrade status` Technology Preview environment variable, otherwise the {oc-first} will not recognize the command and you will not be able to use the feature. 
```

====

.Procedure 

* To enable the `oc adm upgrade status` Technology Preview environment variable, set the `OC_ENABLE_CMD_UPGRADE_STATUS` environmental variable to `true` by running the following command:
+
[source,terminal]
----

> $ OC_ENABLE_CMD_UPGRADE_STATUS=true ./oc adm upgrade status

Reviewer 2: Is the `./` needed? Seems like it's assuming the binary is in current directory, vs running plain `oc` to find the one set in `PATH`:
```suggestion
$ OC_ENABLE_CMD_UPGRADE_STATUS=true oc adm upgrade status
```
For example if I run it on my system with `./`:
```
$ OC_ENABLE_CMD_UPGRADE_STATUS=true ./oc adm upgrade status

zsh: no such file or directory: ./oc
```
vs without `./` (I don't have the newest binary but it at least is trying to run my installed `oc`):
```
$ OC_ENABLE_CMD_UPGRADE_STATUS=true oc adm upgrade status

Unable to connect to the server: dial tcp: lookup api.ci-ln-ypbc3gt-76ef8.aws-2.ci.openshift.org: no such host
```

----
+
[NOTE]
====

> This command sets the environment variable to `true` and then performs the `oc adm upgrade status` command. You do not need to include `$ OC_ENABLE_CMD_UPGRADE_STATUS=true` for future uses of `oc adm upgrade status` once the environment variable has been set.

Reviewer 2: You don't need the `$` prompt when mentioning the variable.

But also I don't think including it as single pass / the one-line form actually sets it globally. If you want to say this, then I believe you'll need to switch to the two-step procedure, with step 1 being:

```
$ export OC_ENABLE_CMD_UPGRADE_STATUS=true
```
and step 2 being:
```
$ oc adm upgrade status
```

The `oc adm upgrade status` command is read-only and will never alter any state in your cluster. 

[IMPORTANT]
====	
The `oc adm upgrade status` command is a Technology Preview feature only. Technology Preview features are not supported with Red Hat production service level agreements (SLAs) and might not be functionally complete. Red Hat does not recommend using them in production. These features provide early access to upcoming product features, enabling customers to test functionality and provide feedback during the development process.

For more information about the support scope of Red Hat Technology Preview features, see link:https://access.redhat.com/support/offerings/techpreview[Technology Preview Features - Scope of Support].
====

> The `oc adm upgrade status` command can be used for clusters from version 4.12 up to the latest supported release. 

Reviewer: I recall we also wanted to mention that to use the feature, you need to enable the TP envvar before running the client, but your cluster does not need to be a TechPreview-enabled cluster.

// Module included in the following assemblies:
//
// * updating/updating_a_cluster/updating-cluster-cli.adoc

:_mod-docs-content-type: PROCEDURE
[id="update-upgrading-oc-adm-upgrade-status_{context}"]
= Retrieving information about a cluster update using oc adm upgrade status (Technology Preview)

When updating your cluster, it is useful to understand how your update is progressing. While the `oc adm upgrade` command returns limited information about the status of your update, this release introduces the `oc adm upgrade status` command as a Technology Preview feature. This command decouples status information from the `oc adm upgrade` command and provides specific information regarding a cluster update, including the status of the control plane and worker node updates.

The `oc adm upgrade status` command is read-only and will never alter any state in your cluster. 

:FeatureName: The `oc adm upgrade status` command
include::snippets/technology-preview.adoc[]

The `oc adm upgrade status` command can be used for clusters from version 4.12 up to the latest supported release. 

[IMPORTANT]
====

> While your cluster does not need to be a Technology Preview-enabled cluster, you must enable the `oc adm upgrade status` Technology Preview environment variable, otherwise the {oc-first} will not recognize the command and you will not be able to use the feature. 

Reviewer: ```suggestion
While your cluster does not need to be a Technology Preview-enabled cluster, you must enable the `OC_ENABLE_CMD_UPGRADE_STATUS` Technology Preview environment variable, otherwise the {oc-first} will not recognize the command and you will not be able to use the feature. 
```

====

.Procedure 

. Set the `OC_ENABLE_CMD_UPGRADE_STATUS` environmental variable to `true` by running the following command:
+
[source,terminal]
----

> $ OC_ENABLE_CMD_UPGRADE_STATUS=true

Reviewer: ```suggestion
$ export OC_ENABLE_CMD_UPGRADE_STATUS=true
```

When done in two steps the variable needs to be exported so that `oc` can read it

----

> +

Reviewer 2: Don't need the `+` continuation here between steps:

```suggestion

```

. Run the `oc adm upgrade status` command:
+
[source,terminal]
----
$ oc adm upgrade status
----
+

> The following output is an example of an update that is progressing successfully:
> +
> .Example output

Reviewer 2: Just a suggestion, but if you wanna trim a line, could rework the sentence into the block title:
```suggestion
.Example output for an update progressing successfully
```
